File: python/mahjong_utils/hora.py
from typing import Optional, Set, List
from typing import Tuple

from pydantic import BaseModel, Field

from mahjong_utils.models.furo import Furo
from mahjong_utils.models.hora_hand_pattern import HoraHandPattern, build_hora_hand
from mahjong_utils.models.tile import Tile
from mahjong_utils.models.wind import Wind
from mahjong_utils.point_by_han_hu import get_parent_point_by_han_hu, get_child_point_by_han_hu
from mahjong_utils.shanten import ShantenResult, shanten
from mahjong_utils.yaku import Yaku
from mahjong_utils.yaku.check import check_yaku


def _calc_han(yaku: Set[Yaku], menzen: bool) -> int:
    if menzen:
        han = sum(map(lambda y: y.han, yaku))
    else:
        han = sum(map(lambda y: y.han - y.furo_loss, yaku))
    return han


class Hora(BaseModel):
    pattern: HoraHandPattern
    dora: int = 0
    extra_yaku: Optional[Set[Yaku]]
    han: int = 0  # 在__init__中计算
    yaku: Set[Yaku] = Field(default_factory=set)  # 在__init__中计算

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self.yaku = check_yaku(self.pattern, self.extra_yaku)
        self.han = _calc_han(self.yaku, self.pattern.menzen)
        if self.han > 0:
            self.han += self.dora

    @property
    def hu(self) -> int:
        return self.pattern.hu

    @property
    def tsumo(self) -> bool:
        return self.pattern.tsumo

    @property
    def self_wind(self) -> Optional[Wind]:
        return self.pattern.self_wind

    @property
    def round_wind(self) -> Optional[Wind]:
        return self.pattern.round_wind

    @property
    def agari(self) -> Tile:
        return self.pattern.agari

    @property
    def parent_point(self) -> Tuple[int, int]:
        """
        亲家和牌点数

        :return: (荣和点数, 自摸各家点数)
        """
        if len(self.yaku) == 0:
            return 0, 0
        elif next(iter(self.yaku)).is_yakuman:
            times = 0
            for yaku in self.yaku:
                times += yaku.han // 13

            ans = get_parent_point_by_han_hu(self.han, self.pattern.hu)
            return ans[0] * times, ans[1] * times
        else:
            return get_parent_point_by_han_hu(self.han, self.pattern.hu)

    @property
    def child_point(self) -> Tuple[int, int, int]:
        """
        子家X番Y符的点数

        :return: (荣和点数, 自摸庄家点数, 自摸闲家点数)
        """
        if len(self.yaku) == 0:
            return 0, 0, 0
        elif next(iter(self.yaku)).is_yakuman:
            times = 0
            for yaku in self.yaku:
                times += yaku.han // 13

            ans = get_child_point_by_han_hu(13, 20)
            return ans[0] * times, ans[1] * times, ans[2] * times
        else:
            return get_child_point_by_han_hu(self.han, self.pattern.hu)


def build_hora(tiles: List[Tile], furo: Optional[List[Furo]], agari: Tile,
               tsumo: bool,
               *,
               dora: int = 0,
               self_wind: Optional[Wind] = None, round_wind: Optional[Wind] = None,
               extra_yaku: Optional[Set[Yaku]] = None) -> Hora:
    """
    根据手牌构造Hora

    :param tiles: 手牌
    :param furo: 副露
    :param agari: 和牌
    :param tsumo: 是否自摸
    :param dora: 宝牌数
    :param self_wind: 自风
    :param round_wind: 场风
    :param extra_yaku: 额外役
    :return: Hora
    """
    if furo is None:
        furo = []

    k = len(tiles) // 3 + len(furo)
    if k != 4:
        raise ValueError("invalid length of _tiles")

    shanten_result = shanten(tiles, furo)
    return build_hora_from_shanten_result(shanten_result, agari, tsumo,
                                          dora=dora, self_wind=self_wind, round_wind=round_wind,
                                          extra_yaku=extra_yaku)


def build_hora_from_shanten_result(
        shanten_result: ShantenResult,
        agari: Tile,
        tsumo: bool,
        *, dora: int = 0,
        self_wind: Optional[Wind] = None, round_wind: Optional[Wind] = None,
        extra_yaku: Optional[Set[Yaku]] = None
) -> Hora:
    """
    根据向听分析结果构造Hora

    :param shanten_result: 向听分析结果
    :param agari: 和牌
    :param tsumo: 是否自摸
    :param dora: 宝牌数
    :param self_wind: 自风
    :param round_wind: 场风
    :param extra_yaku: 额外役
    :return: Hora
    """
    if not shanten_result.with_got:
        raise ValueError("shanten_result without got")
    if shanten_result.shanten != -1:
        raise ValueError("shanten != -1")

    possible_hora = []

    patterns = []

    if shanten_result.regular is not None and shanten_result.regular.shanten == -1:
        patterns += shanten_result.regular.hand.patterns
    if shanten_result.chitoi is not None and shanten_result.chitoi.shanten == -1:
        patterns += shanten_result.chitoi.hand.patterns
    if shanten_result.kokushi is not None and shanten_result.kokushi.shanten == -1:
        patterns += shanten_result.kokushi.hand.patterns

    for pat in patterns:
        for hora_hand in build_hora_hand(pat, agari, tsumo, self_wind, round_wind):
            hora = Hora(pattern=hora_hand, dora=dora, extra_yaku=extra_yaku)
            possible_hora.append(hora)

    best_hora = possible_hora[0]
    for hora in possible_hora:
        if hora.han > best_hora.han or (hora.han == best_hora.han and hora.hu > best_hora.hu):
            best_hora = hora
    return best_hora


__all__ = ("Hora", "build_hora", "build_hora_from_shanten_result")

File: src/commonMain/kotlin/mahjongutils/hora/Hora.kt
@file:OptIn(ExperimentalSerializationApi::class)

package mahjongutils.hora

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import mahjongutils.hanhu.ChildPoint
import mahjongutils.hanhu.ParentPoint
import mahjongutils.hanhu.getChildPointByHanHu
import mahjongutils.hanhu.getParentPointByHanHu
import mahjongutils.models.Furo
import mahjongutils.models.Tile
import mahjongutils.models.Wind
import mahjongutils.models.hand.HoraHandPattern
import mahjongutils.shanten.ShantenResult
import mahjongutils.shanten.ShantenWithGot
import mahjongutils.shanten.shanten
import mahjongutils.yaku.Yaku
import mahjongutils.yaku.Yakus

/**
 * 和牌分析结果
 */
@Serializable
data class Hora internal constructor(
    /**
     * 和牌形
     */
    val pattern: HoraHandPattern,
    /**
     * 宝牌数目
     */
    val dora: Int,
    /**
     * 额外役种
     */
    val extraYaku: Set<Yaku>,
) : HoraInfo by pattern {
    /**
     * 役种
     */
    @EncodeDefault
    val yaku: Set<Yaku> = buildSet {
        if (pattern.menzen) {
            addAll(Yakus.allYakuman.filter { it.check(pattern) })
        } else {
            addAll(Yakus.allYakuman.filter { !it.menzenOnly && it.check(pattern) })
        }
        addAll(extraYaku.filter { it.isYakuman })

        if (isEmpty()) {
            // 非役满情况才判断其他役种
            if (pattern.menzen) {
                addAll(Yakus.allCommonYaku.filter { it.check(pattern) })
            } else {
                addAll(Yakus.allCommonYaku.filter { !it.menzenOnly && it.check(pattern) })
            }
            addAll(extraYaku.filter { !it.isYakuman })
        }
    }

    /**
     * 是否含役满役种
     */
    @EncodeDefault
    val hasYakuman: Boolean = yaku.any { it.isYakuman }

    /**
     * 番
     */
    @EncodeDefault
    val han: Int = run {
        var ans = if (pattern.menzen) {
            yaku.sumOf { it.han }
        } else {
            yaku.sumOf { it.han - it.furoLoss }
        }

        if (ans > 0 && !hasYakuman) {
            ans += dora
        }

        ans
    }

    /**
     * 亲家（庄家）和牌点数
     */
    val parentPoint: ParentPoint
        get() = if (han == 0) {
            ParentPoint(0, 0)
        } else if (hasYakuman) {
            val times = yaku.filter { it.isYakuman }.sumOf { it.han / 13 }
            val ans = getParentPointByHanHu(13, 20)
            ParentPoint(ans.ron * times, ans.tsumo * times)
        } else {
            getParentPointByHanHu(han, hu)
        }

    /**
     * 子家（闲家）和牌点数
     */
    val childPoint: ChildPoint
        get() = if (han == 0) {
            ChildPoint(0, 0, 0)
        } else if (hasYakuman) {
            val times = yaku.filter { it.isYakuman }.sumOf { it.han / 13 }
            val ans = getChildPointByHanHu(13, 20)
            ChildPoint(ans.ron * times, ans.tsumoParent * times, ans.tsumoChild * times)
        } else {
            getChildPointByHanHu(han, hu)
        }
}

/**
 * 和牌分析
 *
 * @param tiles 门前的牌
 * @param furo 副露
 * @param agari 和牌张
 * @param tsumo 是否自摸
 * @param dora 宝牌数目
 * @param selfWind 自风
 * @param roundWind 场风
 * @param extraYaku 额外役种
 * @return 和牌分析结果
 */
fun hora(
    tiles: List<Tile>, furo: List<Furo> = emptyList(), agari: Tile, tsumo: Boolean,
    dora: Int = 0, selfWind: Wind? = null, roundWind: Wind? = null, extraYaku: Set<Yaku> = emptySet()
): Hora {
    val k = tiles.size / 3 + furo.size
    if (k != 4) {
        throw IllegalArgumentException("invalid length of tiles")
    }

    val shantenResult = shanten(tiles, furo, false)
    return hora(shantenResult, agari, tsumo, dora, selfWind, roundWind, extraYaku)

}

/**
 * 和牌分析
 *
 * @param shantenResult 向听分析结果
 * @param agari 和牌张
 * @param tsumo 是否自摸
 * @param dora 宝牌数目
 * @param selfWind 自风
 * @param roundWind 场风
 * @param extraYaku 额外役种
 * @return 和牌分析结果
 */
fun hora(
    shantenResult: ShantenResult, agari: Tile, tsumo: Boolean,
    dora: Int = 0, selfWind: Wind? = null, roundWind: Wind? = null, extraYaku: Set<Yaku> = emptySet()
): Hora {
    if (shantenResult.shantenInfo !is ShantenWithGot) {
        throw IllegalArgumentException("shantenInfo is not with got")
    }
    if (shantenResult.shantenInfo.shantenNum != -1) {
        throw IllegalArgumentException("shantenNum != -1")
    }

    val patterns = buildList {
        if (shantenResult.regular?.shantenInfo?.shantenNum == -1) {
            addAll(shantenResult.regular.hand.patterns)
        }
        if (shantenResult.chitoi?.shantenInfo?.shantenNum == -1) {
            addAll(shantenResult.chitoi.hand.patterns)
        }
        if (shantenResult.kokushi?.shantenInfo?.shantenNum == -1) {
            addAll(shantenResult.kokushi.hand.patterns)
        }
    }

    val possibleHora = patterns.map { pat ->
        HoraHandPattern.build(pat, agari, tsumo, selfWind, roundWind).map { horaHandPat ->
            Hora(horaHandPat, dora, extraYaku)
        }
    }.flatten()
    val hora = possibleHora.maxBy {
        it.han * 1000 + it.pattern.hu  // first key: han, second key: hu
    }
    return hora
}